### ISSUE TITLE
canceled sagas continue executing after task.isRunning() is false and task.done is resolved

### ISSUE THREAD

**Issue by mmerickel** [redux-saga/redux-saga#1242]
Related: https://github.com/redux-saga/redux-saga/issues/1234

Currently there is no API to determine when a task is actually done executing. If redux-saga is providing a mechanism for tasks to perform cleanup then it should provide a way to synchronize that work with the caller.

```js
function* someTask() {
  try {
    while (true) {
      yield call(delay, 1000);
    }
  }
  finally {
    if (yield cancelled()) {
      console.log('task is performing cleanup');
      yield call(delay, 5000);
      console.log('task cleanup complete');
    }
  }
}

function* saga() {
  console.log('starting task');
  const task = yield fork(someTask);
  yield call(delay, 2000);
  yield cancel(task);
  console.log('work continues assuming task is canceled');
}
```

You will see:

```
+0s: starting task
+2s: task is performing cleanup
+4s: work continues assuming task is canceled
+9s: task cleanup complete
```

Similarly `task.isRunning()` is `false` and `task.done` is resolved but in a cancelled state (#1234) so if I wait on it I will be canceled as well. The only workaround right now as mentioned in #408 is to pass my own channel to the task and wait on it. This can be hard to do, especially in the case of `race` and `all` where the library itself may cancel the task. This type of synchronization issue between sagas is a big problem with cleanup and is missing a good story in redux-saga.

Finally, if `someTask` is, itself, waiting on another task to do its own cleanup then the pattern repeats itself and making a new channel per fork is a quite tedious workaround for a clear limitation in the library. Can anyone say that `task.isRunning()` should really be `false` when it's joined to the event loop and able to yield/execute effects?

The correct approach to me is to treat `cancel` as a request which the subtask can respond to by checking the `cancelled` effect. It can then propagate the `cancel` to its subtask or wait for it to complete based on its contract with that task. Finally it will return on its own. Injecting a return into the generator is severely breaking assumptions around the teardown lifecycle of a saga.

I realize the proposed fix above is unlikely to get accepted because it is such a fundamental change to how cancelation works but there does need to be some mechanism for blocking until a task is complete that does not cancel the caller. For example `yield cancel.join(task)` or `yield cancel(task); yield join.doNotPropagateCancel(task)` or something similar.

Possible fixes:

1. `yield cancel(task); yield join(task);` where join would not perform a self-cancelation. This self-cancelation can be done by the caller themselves if they wish. After the join we expect `task.isRunning` to be `false` and `task.isCancelled` to be `true`.

2. `yield cancel.join(task)` which is a special join that waits until the cancel is complete and then returns without any self-cancelation.

I would greatly prefer option 1 but respect the minor incompability issues it presents. Option 2 would be fully backward compatible but redux-saga would be left with weird defaults for join with regard to self-cancelation.

**Comment by Andarist** [comment]
When you have mentioned that joining cancelled task cancels the joiner I have immediately thought that this is a serious bug, but it seems this is done [by design](https://github.com/redux-saga/redux-saga/blob/master/src/internal/proc.js#L212-L225). Just a proof that maintaining some code doesn't mean knowing it all 😄 

Truth to be told I need to think for some time about this issue. Im in the preparation for v1, so it is a perfect moment for some breaking changes, ain't sure though if this is one I would like to do.

That being said I plan to support your use case, just ain't sure about the API yet.

**Comment by mmerickel** [comment]
Well my recommendation for a real breaking change would be to mirror more closely a traditional threading system where you cannot arbitrarily kill threads. In this system `cancel` is just a request, instead of injecting a `return` statement into your generator - and that it is only one level deep. The sagas themselves would then check if they were cancelled using `yield cancelled()` and a new blocking api could be added for them to wait until they are cancelled which could be used in a race with other tasks (imagine `yield race({work: call(delay, 5000), cancel: waitForCancel()}); if (cancel) ...`. This would allow the saga to fully control its lifecycle, cleanup and return when it is ready. It's actually easier in some respects to do work in a thread-based system (ignoring daemonized / detached threads) than redux-saga because you have more guarantees about your lifecycle even though state around you may change at any time. The injection of a `return` at an arbitrary time makes me very worried each time I do a `yield` !

**Comment by Andarist** [comment]
`redux-saga` comes from the different background than a threading systems and it actually conforms to the standard behaviours of joining a cancelled task in fp Task libraries like [folktale's task](http://folktale.origamitower.com/api/v2.0.0/en/folktale.concurrency.html), [fun-task](https://github.com/rpominov/fun-task/), [funfix](https://github.com/funfix/funfix), [avenir](https://github.com/yelouafi/avenir) and more.

However Im very interested in providing APIs to create custom effects and possibly exposing for them internal stuff like `taskQueue`, so other semantics could be implemented easily. Dont know if its even possible (but atm I think it is), dont know when it will be done (I really dont have much time to work on this atm, im really slowly pushing things forward here :/ if you want to collaborate on it, i would be glad to receive any help, introduce you to some internal stuff so the things can be understood easier etc), but for sure I will get to it eventually.

I still need to consider how `task.isRunning()` and `task.done` (which is renamed on master to `task.toPromise()`) behave during cancellation and if the current behaviour is OK.

**Comment by mmerickel** [comment]
> redux-saga comes from the different background than a threading systems and it actually conforms to the standard behaviours of joining a cancelled task in fp Task libraries like folktale's task, fun-task, funfix, avenir and more.

It's fine to have the self-cancellation feature if you feel it's necessary but there really does need to be some way to track a task's completion without the self-cancellation. For example, introduce a `wait` effect or add an option to `toPromise` like `propagateCancel=false`.

**Comment by Andarist** [comment]
As mentioned I want to explore exposing API for custom effects to see if it could support this use case, adding an option to `toPromise` doesn't seem like ideal API and would also make implementation harder to maintain.

**Comment by kennethlynne** [comment]
Just a note for fellow devs: 

just
```
yield cancel(task);
yield take(DONE);
console.log('work continues assuming task is canceled');
```
in saga, and 
```
console.log('task is performing cleanup');
yield call(delay, 5000);
console.log('task cleanup complete');
yield put(DONE);
```
in the worker should do the trick

**Comment by Andarist** [comment]
@kennethlynne while such inter saga communication works, there is no guarantee that `DONE` was dispatched by the same `task`'s instance. Imagine 2 parallel tasks performing the cleanup phase at the same moment, dispatched `DONE` would resume both "joiners"

**Comment by restrry** [comment]
@Andarist @shinima what could be possible action points for this issue? in this discussion were mentioned 2 possible solutions:
- we can create custom effect for asyncCancellation (for instance cancel.resolve(), similarly to `put.resolve()`). that adds  more questions - should we have race that supports async cancellation, for  example.
- it could be resolved in user land as long as we provide users custom effect api.

**Comment by shinima** [comment]
I don't quite follow this issue (It has too much English for me...).   So I just list my understandings, correct me if I'm wrong:

* I believe that `mainTask._isRunning` tracks the running status of the iterator. If we really want to allow users to determine whether a task is actually done executing, we may expose `mainTask` as a public API.
* Customized effects seems a more general solution. However, customized effects would introduce a lot of complexity and we need to consider which internal objects should be exposed to customized effect-runners.  (#1488 is a try to customized effects)
* In my own opinion, asynchronous cancellation logic is not a good idea. We should keep code in finally-block as simple as possible. For example, [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/Events/beforeunload) can be thought as a cancellation logic for a web page. Asynchronous codes (e.g. ajax requests) in `beforeunload` handler won't get executed. Similarly, it is a good idea to only use synchronous and exception-free codes in finally-block.

**Comment by restrry** [comment]
you are right about `mainTask._isRunning`, but exposing just `mainTask` won't solve problem completely, since `toPromise` reflects `task.isRunning` state.

**Comment by Andarist** [comment]
After thinking about it I also don't think async cancellation should be supported, at least not in the core. Don't think there is anything actionable right now for us here, I of course encourage further discussion - would i.e. love to see if other Task-based libraries allow for such patterns.

**Comment by Absvep** [comment]
Maybe someone here knows why `finally` block is always run for me? I can see output of console.log in `finally` always:

`function* loginFlow(username, password) {
  try {
    yield call(loginApi, username, password);
    yield put({ type: LOGIN_SUCCESS });
    yield put({ type: TOGGLE_LOGGED_DONE, payload: true });
    yield put(push('/dashboard'));
  } catch (error) {
    yield put({ type: LOGIN_ERROR, error });
  } finally {
    if (yield cancelled()) {
      console.log('ALWAYS CANCELLED');
      // yield put(replace('/login'));
    }
  }
}

// Watcher saga.
function* loginWatcher() {
  while (true) {
    const { username, password } = yield take(LOGIN_REQUESTING);
    const task = yield fork(loginFlow, username, password);
    const action = yield take([LOGOUT, LOGIN_ERROR]);
    if (action.type === LOGOUT) yield cancel(task);
    yield call(logoutUser);
  }
}`

Here is my question: https://stackoverflow.com/questions/53223958/saga-always-cancelled

**Comment by neurosnap** [comment]
Closing due to inactivity.  If this is still an issue then please reply and I will reopen and try to help.  Thanks!
